WalletWasabi/Services/SatoshiSynchronizer.cs

```diff
@@ -0,0 +1,173 @@
+		{
+			using var ws = new ClientWebSocket();
+			try
+			{
+				await ConnectToSatoshiEndpointAsync(ws).ConfigureAwait(false);
+				await StartReceivingMessagesAsync(ws).ConfigureAwait(false);
+			}
+			catch (WebSocketException)
+			{
+				_eventBus.Publish(new ConnectionStateChanged(false));
+				await Task.Delay(TimeSpan.FromSeconds(5), cancellationToken).ConfigureAwait(false);
+			}
+			catch (Exception e)
+			{
+				Logger.LogError(e);
+			}
+			finally
+			{
+				_eventBus.Publish(new ConnectionStateChanged(false));
+				await DisconnectFromSatoshiEndpointAsync(ws).ConfigureAwait(false);
+			}
+		}
+
+		return;
+
+		async Task ConnectToSatoshiEndpointAsync(ClientWebSocket ws)
+		{
+			ws.Options.Proxy = _socksProxyUri is not null ? new WebProxy(_socksProxyUri) : ws.Options.Proxy;
+			await ws.ConnectAsync(_satoshiEndpointUri, cancellationToken).ConfigureAwait(false);
+
+			_eventBus.Publish(new ConnectionStateChanged(true));
+
+			await WaitForTipHashAsync().ConfigureAwait(false);
+			await HandshakeAsync(localChain.TipHash).ConfigureAwait(false);
+			return;
+
+			async Task WaitForTipHashAsync()
+			{
+				while (localChain.TipHash is null) // Just another hidden communication/synchronization channel
+				{
+					await Task.Delay(100, cancellationToken).ConfigureAwait(false);
+				}
+			}
+
+			async Task HandshakeAsync(uint256 tipHash)
+			{
+				var handshake = new HandshakeMessage(tipHash);
+				await ws.SendAsync(handshake.ToByteArray(), WebSocketMessageType.Binary, true, cancellationToken).ConfigureAwait(false);
+			}
+		}
+
+		async Task DisconnectFromSatoshiEndpointAsync(WebSocket ws)
+		{
+			if (ws.State == WebSocketState.Open)
+			{
+				await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "Closed by client", cancellationToken).ConfigureAwait(false);
+			}
+		}
+
+		async Task StartReceivingMessagesAsync(WebSocket ws)
+		{
+			while (!cancellationToken.IsCancellationRequested)
+			{
+				var buffer = new byte[80 * 1024];
+				var result = await ws.ReceiveAsync(new ArraySegment<byte>(buffer), cancellationToken).ConfigureAwait(false);
+
+				if (result.MessageType == WebSocketMessageType.Close)
+				{
+					return;
+				}
+
+				var messageType = (ResponseMessage)buffer[0];
+				using var reader = new BinaryReader(new MemoryStream(buffer[1..]));
+				switch (messageType)
+				{
+					case ResponseMessage.BlockHeight:
+						var height = reader.ReadUInt32();
+						localChain.SetServerTipHeight(height);
+						_eventBus.Publish(new ServerTipHeightChanged(height));
+						break;
+
+					case ResponseMessage.Filter:
+						var filter = reader.ReadFilterModel();
+						if (localChain.TipHeight + 1 != filter.Header.Height)
+						{
+							Logger.LogError(ChainHeightMismatchError(filter));
+							await RewindAsync(1).ConfigureAwait(false);
+							return;
+						}
+
+						await _bitcoinStore.IndexStore.AddNewFiltersAsync([filter]).ConfigureAwait(false);
+						break;
+
+					case ResponseMessage.HandshakeError:
```

Comment:
In regtest we remove all the filters when the backend starts, that means that a synchronized client knows more filters than the server because the server has to build them all again while the client already knows them all. For that reason the client tries to reconnect and once it does it, it send the bestknownblockchash to the server but the server still doesn't know about it.

This is something that will never happen in testnet or mainnet.